lib/tapioca/runtime/generic_type_registry.rb

```diff
@@ -151,21 +151,30 @@ def create_generic_type(constant, name)
           generic_type
         end
 
-        sig { params(constant: T::Class[T.anything]).returns(T::Class[T.anything]) }
-        def create_safe_subclass(constant)
+        sig { params(constant: T::Class[T.anything], name: String).returns(T::Class[T.anything]) }
+        def create_safe_subclass(constant, name)
           # Lookup the "inherited" class method
           inherited_method = constant.method(:inherited)
           # and the module that defines it
           owner = inherited_method.owner
 
-          # If no one has overriden the inherited method yet, just subclass
+          # If no one has overridden the inherited method yet, just subclass
           return Class.new(constant) if Class == owner
 
+          # Capture this Hash locally, to mutate it in the `inherited` callback below.
+          generic_instances = @generic_instances
+
           begin
             # Otherwise, some inherited method could be preventing us
             # from creating subclasses, so let's override it and rescue
-            owner.send(:define_method, :inherited) do |s|
-              inherited_method.call(s)
+            owner.send(:define_method, :inherited) do |new_subclass|
+              # Register this new subclass ASAP, to prevent re-entry into the `create_safe_subclass` code-path.
+              # This can happen if the sig of the original `.inherited` method references the generic type itself.
+              generic_instances[name] ||= new_subclass
```

Comment:
Admittedly, this is a bit odd.

I would have preferred keeping the simplicity of the "create the thing, then register the thing" as two clean, separate steps, but I don't think that's possible.

Registering the "completed" subclass requires it to first be created, creating it entails running its sigs, which attempts to register it. 🔁

Reply:
You could separate the process in 2 steps. 1. to register the class and 2. to call the original `.inherited`. But this looks good enough 👍 

Should we update the `register_type` comment to add a note about recursiveness?